LineChart and Chart template update

rickimaru18 opened the pull request:
# 📲 What #

* LineChart and Chart template update

# 🤔 Why #

* New chart.

# 🛠 How #

* Updated LineChart.
  - Added LineSeries which controls on how to plot the data in LineChart.
  - Added support to display multiple LineSeries.
  - Added documentation.
* Added support for legends in Chart template.
* In LineChart sample fragment, updated to pass 2 LineSeries to LineChart.
* Added documentations to other files.

# ✅ Acceptance criteria #

- [x] Composable preview.
- [x] Passed Unit Tests
- [x] Passed UI Tests

# ⏰ TODO #

* Unit and UI testing.

# 🔗 References #

* Sample LineChart...
<img width="311" alt="image" src="https://user-images.githubusercontent.com/13028267/113830743-7efa2a00-97b9-11eb-8b81-bcc43923c211.png">
<img width="310" alt="image" src="https://user-images.githubusercontent.com/13028267/113531424-5298af00-95fb-11eb-8fed-49715146c1ec.png">

rickimaru18 committed:
LineChart and Chart template update

# Updated LineChart
- Added LineSeries which controls on how to plot the data in LineChart.
- Added support to display multiple LineSeries.
- Added documentation.

# Updated Chart template
- Added support for legends.

# In LineChart sample fragment
- Updated to pass 2 LineSeries to LineChart.

# Others
- Added documentations.

rexendy-lig commented on kwarchart/src/main/java/com/kwarchart/android/Chart.kt:
```diff
@@ -1,52 +1,182 @@
 package com.kwarchart.android
 
 import androidx.compose.foundation.Canvas
+import androidx.compose.foundation.background
 import androidx.compose.foundation.layout.*
+import androidx.compose.foundation.shape.CircleShape
 import androidx.compose.material.Text
 import androidx.compose.runtime.Composable
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
-import androidx.compose.ui.graphics.Color
 import androidx.compose.ui.graphics.drawscope.DrawScope
 import androidx.compose.ui.graphics.drawscope.scale
 import androidx.compose.ui.tooling.preview.Preview
 import androidx.compose.ui.unit.dp
```
Rick pls ko by group ani
ex.
 import androidx.compose.ui.*

rickimaru18 replied:
OK2. I would just like to ask if this will improve something? Like compile time, different byte code, etc?

rexendy-lig commented on kwarchart/src/main/java/com/kwarchart/android/chart/LineChart.kt:
```diff
@@ -14,109 +14,138 @@ import androidx.compose.ui.tooling.preview.Preview
 import androidx.compose.ui.unit.dp
 import com.kwarchart.android.Chart
 import com.kwarchart.android.ChartCanvas
+import com.kwarchart.android.enum.LegendPosition
+import com.kwarchart.android.enum.LineChartType
 import com.kwarchart.android.model.ChartData
+import com.kwarchart.android.model.Legend
+import com.kwarchart.android.model.LineSeries
 import com.kwarchart.android.util.PathUtils
```
same here

rexendy-lig commented on kwarchart/src/main/java/com/kwarchart/android/util/PathUtils.kt:
```diff
@@ -1,90 +1,69 @@
-            mlX = 0f
-            mlY = 0f
-        }
+//        if (false) {
+//            // first control point
+//            val x1 = p0.x + mlX
+//            val y1 = p0.y + mlY
+//
+//            // second control point
+//            mlX = (p2.x - p0.x) / 2 * smoothness // (lX,lY) is the slope of the reference line
+//            mlY = (p2.y - p0.y) / 2 * smoothness
+//            val x2 = p1.x - mlX
+//            val y2 = p1.y - mlY
+//            return arrayListOf(Offset(x1, y1), Offset(x2, y2))
+//        } else if (false) {
+//            // first control point
+//            val d0 = sqrt((p1.x - p0.x).pow(2f) + (p1.y - p0.y).pow(2f))
+//            val x1 = min(p0.x + mlX * d0, (p0.x + p1.x) / 2)
+//            val y1 = p0.y + mlY * d0
+//
+//            // second control point
+//            val d1 = sqrt((p2.x - p0.x).pow(2f) + (p2.y - p0.y).pow(2f))
+//            mlX = (p2.x - p0.x) / d1 * smoothness
+//            mlY = (p2.y - p0.y) / d1 * smoothness
+//            val x2 = max(p1.x - mlX * d0, (p0.x + p1.x) / 2f)
+//            val y2 = p1.y - mlY * d0
+//
+//            return arrayListOf(Offset(x1, y1), Offset(x2, y2))
+//        }
```
if not needed pls delete

rexendy-lig commented on kwarchart/src/main/java/com/kwarchart/android/Chart.kt:
```diff
@@ -1,52 +1,182 @@
-        modifier = modifier,
-        verticalAlignment = Alignment.CenterVertically
-    ) {
-        Text(
-            text = legendY,
-            color = legendColor,
-            modifier = Modifier.padding(start = 10.dp)
-        )
-        Column(
-            horizontalAlignment = Alignment.CenterHorizontally
+    Column(modifier = modifier) {
+        if (
+            legend != null &&
+            (legend.position == LegendPosition.TOP_LEFT ||
+            legend.position == LegendPosition.TOP ||
```
for fun Chart, pls create a private method pra dli kau taas ang code sa usa ka func and it would be readable
pls check other functions as well, if need pa creatan ug private methods

rexendy-lig reviewed (changes requested):
Be consistent in using {} for if statments..ok ra if wlay {} if one line lng ang sulod sa {}
add comments sa mga codes if needed

thank you Rick

rickimaru18 committed:
Legend update and added LineChart's axes values

# In Chart template
- Segragate legends into different composables.

# In LineChart
- Axes values are now displayed.
- Use canvas' normal coordinates to properly display drawn texts.

rexendy reviewed (commented):
@rickimaru18  ang tanan padding and weight, mausab2x pani? if dli kay buhati lng constant var..salamatzz

rickimaru18 commented:
> @rickimaru18 ang tanan padding and weight, mausab2x pani? if dli kay buhati lng constant var..salamatzz

For the weight, no. 
For the padding, these properties can be controlled by the user in the future.

rickimaru18 committed:
Initial test setup and create constant variables for weights

# Added initial unit and UI test
- UI test for LineChart.
- Unit test for PathUtils.

# Update versions of gradle and compose

# Create constant variables for weights in Chart and LineChart

rickimaru18 committed:
Rename LineChart test file and create chart utility funcitonn

# Renamed LineChartTest to LineChartInstrumentedTest

# Added new utility function for chart axes values

# In LineChart Y-axis values
- Use newly added utility function, "getAxisValues(...)".

# In unit tests
- Added unit test for ChartUtils.
- Reverse assertion values in PathUtilsUnitTest.

rickimaru18 committed:
Display axes names in sample LineChart fragment and UI test update

# In LineFragment, set X and Y axes names

# In LineChartInstrumentedTest, added axes names test

rexendy commented:
nice @rickimaru18 
pls merge using git flow

thanks
